### PR TITLE
Modularization for AWS terraform scripts

### LINKED ISSUE #628: AWS terraform default to multiple environments
Our terraform should default to a setup in which one can make multiple instantiations (environments). One just needs to use it as a module instead of changing the terraform directly. The layout change is outlined here, along with some additional discussion:

https://github.com/pingcap/tidb-operator/pull/616#issuecomment-507914194

### PR DESCRIPTION
**Pull request by aylei**
### What problem does this PR solve? 

close #628 

### What is changed and how does it work?

As the issue stated, this PR separate the modules of AWS terraform scripts more clearly and suggest the best practice to manage multiple environments. 

The directory name and UX of `aws` is not changed in favor of consistence, but it should be easy for uses to create a new directory and write some terraform scripts from scratch following the  README.

This PR also introduce two minor features based on the community feedbacks:

- Now bastion machine can be configured to ssh to worker nodes, and this is enabled by default;
- tidb-cluster can be configured to not create the real helm release (means only provision worker nodes), which is useful if users want to manage helm on their own;

### Check List 

Tests 

Not yet

Side effects

 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
   - [ ] docs-cn
   - [ ] docs
### Does this PR introduce a user-facing change?:
 
No

### CODE REVIEW

File: deploy/modules/share/tidb-cluster-release/main.tf [hunk 1]
```diff
@@ -1,5 +1,5 @@
 resource "null_resource" "wait-tiller-ready" {
```
Comment (aylei) [inline deploy/modules/share/tidb-cluster-release/main.tf]: This module is separated in favor of reusing across cloud providers

**Comment by gregwebs** [comment]
Looks like there are some conflicts.

**Comment by gregwebs** [comment]
This looks great! However, I think putting this in a separate advanced section will end up with poor results. Once someone edits the module directly they will have multiple problems
* It will be difficult to start a new environment because direct changes have already been made to the module.
* It is very difficult (I don't feel comfortable doing this, but in theory it is possible with terraform import) to transition that existing setup into an environment setup

Consider direct edits to be like distributing a python library and telling users to edit constants in the source code rather than using it as a library and making function calls.

Its a lot better to make a clean break from the past since we just created this and only have users using it in a testing phase right now.

**Comment by gregwebs** [comment]
Following on the above comment, the way to make it easy to use it to start with using as modules is to make an existing environment called "default" that is focused on ease of use rather than flexibility.

**Comment by gregwebs** [comment]
The default environment shouldn't need to be much more than

```
module "default" {
  source = "../"
}
```

**Comment by aylei** [comment]
@gregwebs Thanks for your thoughtful suggestions!

The reason why I keep the `deploy/aws` directory "as is" is to not break users.

And I think we have to reach a consensus about the purpose of the `deploy/aws` directory.

In this PR, I consider it as an example of how to compose the reusable modules located in the `deploy/modules` directory rather than a reusable module itself. Because It is hard to write a single module that provisions multiple tidb-clusters.

It is possible to write a reusable module that provisions VPC, bastion, key-pair and EKS, which effectively composes these four modules. This will reduce the code length to start a new environment, but I suppose composing these modules should require little effort just as the README stated.

So, the first problem you mentioned is now addressed: users don't call the `deploy/aws` module, they call the modules in the `deploy/modules` dir instead.

Another problem you mentioned is transition, actually they don't need, because the `deploy/aws` is already an individual environment. It is parallel to the dirs created by users, e.g. `deploy/aws-staging`, `deploy/aws-production`.

**Comment by aylei** [comment]
> The default environment shouldn't need to be much more than
> 
> ```
> module "default" {
>   source = "../"
> }
> ```

IIUC, this example is a different UX for users. There are two UXs that may make sense:

1. provide a "big", highly customizable module,  users call this module like the code example above;
2. provide several customizable small modules as shown in the PR, users call these modules like the code example in the `README.md`

I suppose both of them has pros and cons.

Comparing to UX-2, UX-1 has:

* Pros:
    * Easy to use
    * Less error-prone because of proper encapsulation
* Cons:
    * Inflexible, all the customization possibilities should be exposed via module variables
    * (things I concern) requires too much developing efforts than we can afford

Actually I don't bias on either method technically, but I do prefer the UX-2 in favor of the overall ROI.

**Comment by aylei** [comment]
@tennix Could you please take a look?

**Comment by aylei** [comment]
> However, I think putting this in a separate advanced section will end up with poor results. Once someone edits the module directly they will have multiple problems.

Yes, I agree we should emphasize this section for the users who need multiple environment management. I will try to refine this README

**Comment by aylei** [comment]
Having said that, I will try to investigate whether I can build a monolithic reusable module in an acceptable amount of development

**Comment by tennix** [comment]
The quick-start guide should be easy enough, so users can really quick start. In most cases, this setup would be for testing. After they get a glimpse of what it looks like, they will try to deploy a production environment. And the quick-start setup may be wiped out. So I agree with @aylei to keep the quick-start as it is and document the best practice guide of how to manage production environment.

**Comment by gregwebs** [comment]
okay, I get your point, and I agree with it. Things are well organized now.
But I would still move what is in `deploy/aws` into an environment directory. I think this update will already break existing users.

**Comment by aylei** [comment]
@gregwebs I've addressed some merge conflicts, could you please take a look again?

And, I've investigate if we can approach to the structure you suggested, and I find that it is possible in the future! Once the https://github.com/hashicorp/terraform/issues/17179 addressed, we can define multiple tidb clusters by a single map variable and things won't break when some of the clusters are removed (which is impossible now because currently the loop of resources is index based, which means we cannot remove element in the middle of the list).

**Comment by aylei** [comment]
@tennix PTAL

File: deploy/modules/share/tidb-cluster-release/data.tf [hunk 2]
```diff
@@ -0,0 +1,11 @@
+data "external" "tidb_hostname" {
+  depends_on  = [helm_release.tidb-cluster]
+  working_dir = path.cwd
+  program     = ["bash", "-c", "kubectl --kubeconfig ${var.kubeconfig_filename} get svc -n ${var.cluster_name} ${var.cluster_name}-tidb -o json | jq '.status.loadBalancer.ingress[0]'"]
```
Comment (tennix) [inline deploy/modules/share/tidb-cluster-release/data.tf]: Can this read the right kubeconfig file when in a separated module?

File: deploy/modules/aws/tidb-cluster/cluster.tf [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  cluster_version            = var.cluster_version
+  pd_count                   = var.pd_count
+  tikv_count                 = var.tikv_count
+  tidb_count                 = var.tidb_count
+  tidb_cluster_chart_version = var.tidb_cluster_chart_version
+  override_values            = var.override_values
+  local_exec_interpreter     = var.local_exec_interpreter
+  base_values                = file("${path.module}/values/default.yaml")
+  kubeconfig_filename        = var.eks.kubeconfig_filename
+}
```
Comment (tennix) [inline deploy/modules/aws/tidb-cluster/cluster.tf]: No newline at the end of file.

File: deploy/modules/aws/tidb-cluster/cluster.tf [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  source = "../../share/tidb-cluster-release"
+
+  cluster_name               = var.cluster_name
+  cluster_version            = var.cluster_version
+  pd_count                   = var.pd_count
+  tikv_count                 = var.tikv_count
+  tidb_count                 = var.tidb_count
+  tidb_cluster_chart_version = var.tidb_cluster_chart_version
+  override_values            = var.override_values
+  local_exec_interpreter     = var.local_exec_interpreter
```
Comment (tennix) [inline deploy/modules/aws/tidb-cluster/cluster.tf]: Can this work if a user does not provide `local_exec_interpreter`?

File: deploy/modules/aws/bastion/variables.tf [hunk 5]
```diff
@@ -0,0 +1,38 @@
+variable "bastion_name" {
+  description = "Name of the EKS cluster. Also used as a prefix in names of related resources."
+  type = string
+}
+
+variable "enable_ssh_to_workers" {
+  description = "Whether enable ssh from bastion to workers, if true, the worker_security_group_id must be provided"
+  default = false
```
Comment (tennix) [inline deploy/modules/aws/bastion/variables.tf]: Can we enable this by default?

File: deploy/modules/aws/bastion/bastion.tf [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  instance_count              = 1
+  ami                         = data.aws_ami.amazon-linux-2.id
+  instance_type               = var.bastion_instance_type
+  key_name                    = var.key_name
+  associate_public_ip_address = true
+  monitoring                  = false
+  user_data                   = file("${path.module}/bastion-userdata")
+  vpc_security_group_ids      = [aws_security_group.accept_ssh_from_local.id]
+  subnet_ids = var.public_subnets
+}
```
Comment (tennix) [inline deploy/modules/aws/bastion/bastion.tf]: No newline at end of file

File: deploy/.gitignore [hunk 7]
```diff
@@ -0,0 +1,6 @@
+.terraform/
+credentials/
+terraform.tfstate
+terraform.tfstate.backup
+.terraform.tfstate.lock.info
+kubeconfig_*.yaml
```
Comment (tennix) [inline deploy/.gitignore]: No newline at end of file

File: deploy/modules/aws/tidb-operator/README.md [hunk 8]
```diff
@@ -2,6 +2,5 @@ The `tidb-operator` module for AWS spins up a control plane for TiDB in Kubernet
 
 - An EKS cluster
 - A auto scaling group to run the control pods listed below
-- TiDB operator, including `tidb-controller-manager` and `tidb-scheduler`
```
[the remark this answers is not in the document]
Comment (aylei) [reply deploy/modules/aws/tidb-operator/README.md]: This is removed by accident

File: deploy/modules/share/tidb-cluster-release/data.tf (hunk 2 above)
Comment (aylei) [reply deploy/modules/share/tidb-cluster-release/data.tf]: Yes, the kubeconfig_filename should be the relative path to `path.cwd` by convention, and this is now guaranteed by all the modules we provide

File: deploy/modules/aws/tidb-cluster/cluster.tf (hunk 4 above)
Comment (aylei) [reply deploy/modules/aws/tidb-cluster/cluster.tf]: Yes, there is a reasonable default `["bash", "-c"]`

File: deploy/modules/aws/tidb-cluster/variables.tf [hunk 9]
```diff
@@ -1,3 +1,72 @@
...
+  type    = string
+  default = "c5.4xlarge"
+}
+
+variable "monitor_instance_type" {
+  type    = string
+  default = "c5.2xlarge"
+}
+
+variable "override_values" {
```
[the remark this answers is not in the document]
Comment (aylei) [reply deploy/modules/aws/tidb-cluster/variables.tf]: Okay, this is the content of `values.yaml`, it is recommend to provide the content by the `file()` function call from a local file

File: deploy/modules/aws/bastion/variables.tf (hunk 5 above)
Comment (aylei) [reply deploy/modules/aws/bastion/variables.tf]: No, the `worker_security_group_id` must be provided with this enabled, which can not have a default value. So user will be forced to provide the id or disable this explicitly if we enable this by default.

**Comment by aylei** [comment]
@tennix I've addressed some of the review comments and add comments for others, PTAL again

**Comment by aylei** [comment]
@gregwebs I've addressed some comments from tennix, could you please take a look again (again🤣)? Thanks!

**Pull request merged**